File: Sources/Buckett/Services/BucketAliases.swift
import Foundation

/// User-defined display names for buckets. Aliases are purely cosmetic — every
/// API request still uses the real bucket name. Keyed by "accountUUID|bucket".
@MainActor
final class BucketAliases: ObservableObject {
    static let shared = BucketAliases()

    @Published private(set) var aliases: [String: String] = [:]

    private var fileURL: URL {
        AccountStore.supportDirectory.appendingPathComponent("bucket-aliases.json")
    }

    private init() {
        if let data = try? Data(contentsOf: fileURL),
           let decoded = try? JSONDecoder().decode([String: String].self, from: data) {
            aliases = decoded
        }
    }

    private static func key(accountID: UUID, bucket: String) -> String {
        "\(accountID.uuidString)|\(bucket)"
    }

    func alias(accountID: UUID, bucket: String) -> String? {
        aliases[Self.key(accountID: accountID, bucket: bucket)]
    }

    /// Alias when set, otherwise the real bucket name.
    func displayName(accountID: UUID, bucket: String) -> String {
        alias(accountID: accountID, bucket: bucket) ?? bucket
    }

    /// Sets (or clears with nil/empty) the alias for a bucket.
    func setAlias(_ alias: String?, accountID: UUID, bucket: String) {
        let key = Self.key(accountID: accountID, bucket: bucket)
        let trimmed = alias?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        if trimmed.isEmpty {
            aliases.removeValue(forKey: key)
        } else {
            aliases[key] = trimmed
        }
        save()
    }

    private func save() {
        try? FileManager.default.createDirectory(
            at: AccountStore.supportDirectory, withIntermediateDirectories: true
        )
        if let data = try? JSONEncoder().encode(aliases) {
            try? data.write(to: fileURL, options: .atomic)
        }
    }
}

File: windows/src/Buckett/Services/BucketAliases.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.Json;
using CommunityToolkit.Mvvm.ComponentModel;

namespace Buckett.Services;

/// User-defined display names for buckets. Aliases are purely cosmetic — every
/// API request still uses the real bucket name. Keyed by "accountUUID|bucket".
public sealed class BucketAliases : ObservableObject
{
    public static BucketAliases Shared { get; } = new();

    private Dictionary<string, string> _aliases = new();

    public IReadOnlyDictionary<string, string> Aliases => _aliases;

    private static string FilePath => Path.Combine(AppPaths.SupportDirectory, "bucket-aliases.json");

    private BucketAliases()
    {
        try
        {
            if (File.Exists(FilePath))
            {
                var json = File.ReadAllText(FilePath);
                var decoded = JsonSerializer.Deserialize<Dictionary<string, string>>(json);
                if (decoded != null) _aliases = decoded;
            }
        }
        catch (Exception error)
        {
            Log.Warn($"failed to load bucket aliases: {error.Message}");
        }
    }

    private static string Key(Guid accountID, string bucket) =>
        $"{accountID.ToString("D").ToUpperInvariant()}|{bucket}";

    public string? Alias(Guid accountID, string bucket) =>
        _aliases.TryGetValue(Key(accountID, bucket), out var alias) ? alias : null;

    /// Alias when set, otherwise the real bucket name.
    public string DisplayName(Guid accountID, string bucket) => Alias(accountID, bucket) ?? bucket;

    /// Sets (or clears with null/empty) the alias for a bucket.
    public void SetAlias(string? alias, Guid accountID, string bucket)
    {
        var key = Key(accountID, bucket);
        var trimmed = alias?.Trim() ?? "";
        if (trimmed.Length == 0)
        {
            _aliases.Remove(key);
        }
        else
        {
            _aliases[key] = trimmed;
        }
        Save();
        OnPropertyChanged(nameof(Aliases));
    }

    private void Save()
    {
        try
        {
            AppPaths.EnsureSupportDirectory();
            File.WriteAllText(FilePath, JsonSerializer.Serialize(_aliases));
        }
        catch (Exception error)
        {
            Log.Warn($"failed to save bucket aliases: {error.Message}");
        }
    }
}
